src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java

```diff
@@ -0,0 +1,112 @@
+package io.vertx.cassandra.impl;
+
+import com.datastax.driver.core.ResultSet;
+import io.vertx.cassandra.CassandraRowStream;
+import io.vertx.cassandra.Row;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class CassandraRowStreamImpl implements CassandraRowStream {
+
+  private final ResultSet datastaxResultSet;
+  private final Vertx vertx;
+  private final Iterator<com.datastax.driver.core.Row> resultSetIterator;
+  private final Lock iteratorAccessLock = new ReentrantLock();
+  private final Context context;
+
+  private volatile Handler<Throwable> exceptionHandler;
+  private volatile Handler<Row> rowHandler;
+  private volatile Handler<Void> endHandler;
+
+  private AtomicBoolean paused = new AtomicBoolean(false);
+
+  public CassandraRowStreamImpl(ResultSet result, Vertx vertx) {
+    this.vertx = vertx;
+    datastaxResultSet = result;
+    resultSetIterator = result.iterator();
+    context = vertx.getOrCreateContext();
+  }
+
+  @Override
+  public CassandraRowStream exceptionHandler(Handler<Throwable> handler) {
+    exceptionHandler = handler;
+    return this;
+  }
+
+  @Override
+  public CassandraRowStream handler(Handler<Row> handler) {
+    rowHandler = handler;
+    if (handler != null) {
+      fireStream();
+    }
+    return this;
+  }
+
+  @Override
+  public CassandraRowStream pause() {
+    paused.set(true);
+    return this;
+  }
+
+  @Override
+  public CassandraRowStream resume() {
+    paused.set(false);
+    fireStream();
+    return this;
+  }
+
+  @Override
+  public CassandraRowStream endHandler(Handler<Void> handler) {
+    endHandler = handler;
+    return this;
+  }
+
+  private void fireStream() {
+    if (!paused.get() && rowHandler != null) {
+      int availableWithoutFetching = datastaxResultSet.getAvailableWithoutFetching();
+      for (int i = 0; i < availableWithoutFetching; i++) {
+
+        iteratorAccessLock.lock();
```

Comment:
When you use a lock, make sure it is unlocked with this pattern:

```java
lock.lock()
try {
  // do stuff
} finally {
  lock.unlock()
}
```

Reply:
But anyway, it is better to stick to plain synchronization as much as possible. Because when the code will be used on an event loop, the JVM will optimize it with biased locking.

Reply:
Done